platform/source/mbed_retarget.cpp

```diff
@@ -756,11 +756,13 @@ extern "C" int PREFIX(_write)(FILEHANDLE fh, const unsigned char *buffer, unsign
 extern "C" ssize_t write(int fildes, const void *buf, size_t length)
 {
 #if MBED_CONF_PLATFORM_STDIO_MINIMAL_CONSOLE_ONLY
+// Microlib does not pass STDOUT_FILENO or STDERR_FILENO when _sys_write is called.
+#if !defined(__MICROLIB)
```

Comment:
If this is about microlib passing weird things to _sys_write, then the logical fix is to put the workaround in _sys_write. Make it call `write` with STDOUT_FILENO. You don't want to disable error checking for other users calling `write`.

Reply:
microlib first opens `:tt` for which our `open()` in `mbed_retarget` returns 0 (stdin):
https://github.com/ARMmbed/mbed-os/blob/83c1b347121863277c6e4305f19dc3b279abac3a/platform/source/mbed_retarget.cpp#L555-L563

The `open()` code suggests `:tt` should (?) be opened three times (once for in, out, err each) but debugger shows it's only opened once. This might be related to why `write` was called with `0` file descriptor instead of `1` (stdout)? @kjbracey-arm

Reply:
> ...then the logical fix is to put the workaround in _sys_write.

The current change seemed more logical to me. If the postman is scared of dogs and you have a dog to protect your home, you keep your dog away from the door when the postman arrives. You don't ask for a new postman...🌝.

> You don't want to disable error checking for other users calling write.

How will other users be affected since the check is present for everything except Microlib (`#if !defined(__MICROLIB)`)?

Reply:
`write` is a public function. People can use it for their own files directly (`open`/`read`/`close`), regardless of any issues in microlib.

In the minimal console case, you do want to trap someone doing something weird - given the above call sequence you want to give an error when given the "-1" handle that `open` gave you to open a file, not send the stuff that should have gone to file to the console. That check shouldn't be removed because of an issue in one user.

At a minimum you'd want to at least ensure it was a valid TTY handle of some sort - maybe accept `<= STDERR_FILENO` if the numbers from microlib are fuzzy.

Now, LDong's description looks on the point - something weird is happening with the way microlib opens our streams. Is it that microlib doesn't actually open 3 in sequence like the code assumes? Is it opening on-demand - first call to each of stdin/stdout/stderr?

Does it at least open twice, so we could look at the mode and return either `STDIN_FILENO` or `STDOUT_FILENO`?

Reply:
> At a minimum you'd want to at least ensure it was a valid TTY handle of some sort - maybe accept `<= STDERR_FILENO` if the numbers from microlib are fuzzy.

This is what `mbed_file_handle()` does (when not using minimal-console):
https://github.com/ARMmbed/mbed-os/blob/83c1b347121863277c6e4305f19dc3b279abac3a/platform/source/mbed_retarget.cpp#L371-L386
when the file descriptor is < 3, it just uses `get_console()` which returns the _default console_ in most cases.  (Also the comment there suggests descriptors 0,1,2 indeed may not be open by the C lib.)

Reply:
There are a couple of paths into all this code. You are allowed to have an `mbed_override_console` that sends stdin, stdout and stderr to 3 separate `FileHandle` devices - that's probably the only place we really care. If `mbed_override_console` isn't added by the app or target, then all 3 do go to the same device anyway.

Having a restriction that `mbed_override_console` doesn't allow different streams for microlib because it doesn't properly distinguish wouldn't be a huge deal. But as I said in my previous comment - maybe it's only not distinguishing because we're not trying. Those `:tt` previously would have been us knowing what name it was going to give us regardless. But maybe it's now actually obeying them, and we could change them.

Reply:
`_sys_open()` is called once with `name: ":tt"` and `openflags: 4 (O_RDONLY | O_NONBLOCK)` . Does it make sense to return `STDOUT_FILENO` with these values...🤔

Reply:
4 is `OPEN_W` for the ARM C library (see `ARMCompiler6.13/include/rt_sys.h`):

```C
/*
 * Open a file. May return -1 if the file failed to open.
 */
extern FILEHANDLE _sys_open(const char * /*name*/, int /*openmode*/);
/*
 * openmode is a bitmap, whose bits are given below. They
 * correspond directly to the ANSI mode specification.
 */
#define OPEN_R 0
#define OPEN_W 4
#define OPEN_A 8
#define OPEN_B 1
#define OPEN_PLUS 2

/*
 * These names should be special strings which will be recognised
 * by _sys_open and will cause it to return the standard I/O
 * handles, instead of opening a real file.
 */
extern const char __stdin_name[];
extern const char __stdout_name[];
extern const char __stderr_name[];
```

That may be because your app is only using `stdout` though.

What if you use stdin, or stderr, either before or after?